### PR TITLE
providers.Interface: rename ValidateDataSourceConfig to ValidateDataResourceConfig

### PR DESCRIPTION
**Pull request by mildwonkey**
I split this into 2 commits to isolate changes to the interface from changes to the protocol, though in the end I don't think it makes a useful difference for reviewers (ymmv). 

@paddycarver I especially want to bring this to your attention, since it differs from the name change we agreed upon (sort of; we've added a new change). 

Martin and I discussed why `ValidateResourceTypeConfig` was named the way it was to start and in the end decided that it's fine as is, but now `ValidateDataSource` config looked wrong, so we settled on `ValidateDataResourceConfig` as a compromise. (I might be more accurate to name them `ValidateManagedResourceConfig` and `ValidateDataResourceConfig`, but also we practically never fully spell out "managed resource", just using "resource" in all cases, so 🤷 )

We also discussed renaming `NestedType`, as you had asked about. I still feel like its clearer as-is, since `NestedType` is a direct alternative to `Type` (and again, if we were starting over it'd ALL be different); I'm curious if that explanation helps it make more sense to you, or not. (It doesn't matter how much sense it makes to me, I'm not writing a provider anytime soon)

I am going to block this PR on https://github.com/hashicorp/terraform/pull/27873 so I don't make Martin deal with the merge conflicts 😀

### CODE REVIEW

**Comment by paddycarver** [comment]
> Martin and I discussed why ValidateResourceTypeConfig was named the way it was to start and in the end decided that it's fine as is

I'd be very interested to hear that backstory.

I'm also unclear if this means that you're going with `ValidateResourceTypeConfig` or `ValidateResourceConfig` as the final result.

> I'm curious if that explanation helps it make more sense to you, or not.

It doesn't, really, but I consider it such a minor issue that I'm definitely not going to push too hard on it. I don't think most provider developers will ever be exposed to it. I'd be interested in chatting sometime about it, though, so we can make sure we're surfacing the same semantics and ideas in the SDK that you're surfacing in the protocol, and that if we're changing words, we still are capturing intent.

**Comment by mildwonkey** [comment]
Ah, I'm sorry, I hadn't finished typing out that thought. 

A managed resource is the actual, provisioned (or managed) resource - `aws_instance.foo` - while a managed resource type is basically the schema - the `aws_instance` part. These functions work with configuration, not real resources -  they could in principle be operating on a totally hypothetical configuration - so that's why we had originally called that `ValidateResourceTypeConfig`. 

Similarly (though much less consistently), we differentiate between data resources types (frequently just data sources) and data resource instances. 

Pragmatically, this distinction may only matter to core developers, which is why we went with your request for `ValidateResourceConfig`. `ValidateDataSourceConfig` was renamed `ValidateDataResourceConfig` so those two functions were consistent with each other. 

I like how Martin described it better, so I'll copy his explanation to me here:
> "data source" is to "resource type" as "data resource" is to "managed resource", with "data source" just kinda being a vestigial shorter way of saying "data resource type"

File: builtin/providers/terraform/provider.go [hunk 1]
```diff
@@ -40,11 +40,11 @@ func (p *Provider) ValidateProviderConfig(req providers.ValidateProviderConfigRe
 }
 
 // ValidateDataSourceConfig is used to validate the data source configuration values.
-func (p *Provider) ValidateDataSourceConfig(req providers.ValidateDataSourceConfigRequest) providers.ValidateDataSourceConfigResponse {
+func (p *Provider) ValidateDataResourceConfig(req providers.ValidateDataResourceConfigRequest) providers.ValidateDataResourceConfigResponse {
```
Comment (alisdair) [inline builtin/providers/terraform/provider.go:43]: Can we update this comment too?

**Review by alisdair — APPROVED** [review]
Changes look good to me! I spotted one comment still referring to `ValidateDataSourceConfig`

File: builtin/providers/terraform/provider.go (hunk 1 above)
Comment (mildwonkey) [reply builtin/providers/terraform/provider.go:43]: ty!

**Comment by paddycarver** [comment]
> Ah, I'm sorry, I hadn't finished typing out that thought.
> 
> A managed resource is the actual, provisioned (or managed) resource - `aws_instance.foo` - while a managed resource type is basically the schema - the `aws_instance` part. These functions work with configuration, not real resources - they could in principle be operating on a totally hypothetical configuration - so that's why we had originally called that `ValidateResourceTypeConfig`.
> 
> Similarly (though much less consistently), we differentiate between data resources types (frequently just data sources) and data resource instances.
> 
> Pragmatically, this distinction may only matter to core developers, which is why we went with your request for `ValidateResourceConfig`. `ValidateDataSourceConfig` was renamed `ValidateDataResourceConfig` so those two functions were consistent with each other.
> 
> I like how Martin described it better, so I'll copy his explanation to me here:
> 
> > "data source" is to "resource type" as "data resource" is to "managed resource", with "data source" just kinda being a vestigial shorter way of saying "data resource type"

Ah, this is super interesting and full of information I'd never heard before!

I think most provider developers don't distinguish between Resource Types and Resource Instances that much, but I could be wrong. We usually talk about schemas vs. state/config, but they all belong to "a resource".

I think I've spent more of your time on this particular phrasing than is remotely warranted by the amount of impact it will have, though. I appreciate the background and information, though!

**Pull request merged**